File: app/src/main/java/com/xpf/android/performance/optimization/MainActivity.java
package com.xpf.android.performance.optimization;

import android.os.Bundle;
import android.view.ViewTreeObserver;
import android.widget.Button;
import android.widget.ImageView;

import androidx.appcompat.app.AppCompatActivity;

import com.xpf.android.performance.optimization.utils.LaunchRecord;

public class MainActivity extends AppCompatActivity {

    private Button btnTest;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btnTest = findViewById(R.id.btnTest);

        btnTest.setOnClickListener(v -> generateImageViews());

        btnTest.getViewTreeObserver().addOnDrawListener(() ->
                LaunchRecord.INSTANCE.end("onDraw")
        );
    }

    private void generateImageViews() {
        for (int i = 0; i < 1000; i++) {
            ImageView imageView = new ImageView(this);
        }
        // 手动 gc
        System.gc();
        // 强制回收 Finalizer 对象
        System.runFinalization();
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        LaunchRecord.INSTANCE.end("onWindowFocusChanged");
    }
}

File: app/src/main/java/com/xpf/android/performance/optimization/MainActivity.kt
package com.xpf.android.performance.optimization

import android.os.Bundle
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import com.xpf.android.performance.optimization.databinding.ActivityMainBinding
import com.xpf.android.performance.optimization.utils.LaunchRecord.end

class MainActivity : AppCompatActivity() {

    private val binding: ActivityMainBinding by lazy {
        ActivityMainBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        binding.btnTest.setOnClickListener { generateImageViews() }
        binding.btnTest.viewTreeObserver.addOnDrawListener { end("onDraw") }
    }

    private fun generateImageViews() {
        for (i in 0..999) {
            val imageView = ImageView(this)
        }
        // 手动 gc
        System.gc()
        // 强制回收 Finalizer 对象
        System.runFinalization()
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        end("onWindowFocusChanged")
    }
}